apps_download/models/product_product.py

```diff
@@ -0,0 +1,107 @@
+# -*- coding: utf-8 -*-
+
+import os
+import tempfile
+import shutil
+import logging
+import base64
+import subprocess
+import time
+from odoo.exceptions import ValidationError
+from odoo import models, fields, api, _
+_logger = logging.getLogger(__name__)
+
+
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
+
+    dependent_product_ids = fields.Many2many(
+        'product.product', 'prto_validateoduct_dependent_rel',
+        'src_id', 'dest_id', string='Dependent Products'
+    )
+    module_path = fields.Char('Module Path')
+
+    @api.constrains('dependent_product_ids')
+    def check_dependent_recursion(self):
+        for product in self:
+            def child_dependency_check(product_dependent_ids, children):
+                res = self.env['product.product']
+                for child in children:
+                    if not child.dependent_product_ids:
+                        continue
+                    if child in product_dependent_ids:
+                        raise ValidationError(
+                            _('Error: You cannot create recursive dependency.')
+                        )
+                    product_dependent_ids += child
+                    child_dependency_check(product_dependent_ids,
+                                           child.dependent_product_ids)
+                return res
+
+            child_dependency_check(product, product.dependent_product_ids)
+
+    @api.multi
+    def create_dependency_list(self):
+        ret_val = {}
+
+        def child_dependency(children):
+            res = self.env['product.product']
+            for child in children:
+                if not child.dependent_product_ids:
+                    continue
+                res += child.dependent_product_ids
+                res += child_dependency(child.dependent_product_ids)
+            return res
+        for product in self:
+            ret_val[product.id] = product.dependent_product_ids
+            if product.dependent_product_ids:
+                ret_val[product.id] += child_dependency(
+                    product.dependent_product_ids)
+        return ret_val
+
+    @api.multi
+    def generate_zip_file(self):
+        for product in self:
+            if not product.module_path:
+                continue
+            tmpdir = tempfile.mkdtemp()
+            tmpdir2 = tempfile.mkdtemp()
+            dependent_product_ids = product.create_dependency_list(
```

Comment:
Please perform the method call on one line and the subsequent assignment on the other. Breaking the method call into two lines, then performing a dict lookup afterwards is hard to follow IMO. Also, the variable should not have the `ids` suffix, because it is a recordset.

```
dependent_products = product.create_dependency_list()
dependent_products = dependent_products[product.id]
```